src/main/java/io/reactivex/internal/operators/flowable/FlowableTimeout.java

```diff
@@ -48,299 +47,343 @@ public FlowableTimeout(
+            s.onSubscribe(parent);
+            parent.startFirstTimeout(firstTimeoutIndicator);
+            source.subscribe(parent);
         }
     }
 
-    static final class TimeoutSubscriber<T, U, V> implements FlowableSubscriber<T>, Subscription, OnTimeout {
+    interface TimeoutSelectorSupport extends TimeoutSupport {
+        void onTimeoutError(long idx, Throwable ex);
+    }
+
+    static final class TimeoutSubscriber<T> extends AtomicLong
+    implements FlowableSubscriber<T>, Subscription, TimeoutSelectorSupport {
+
+        private static final long serialVersionUID = 3764492702657003550L;
+
         final Subscriber<? super T> actual;
-        final Publisher<U> firstTimeoutIndicator;
-        final Function<? super T, ? extends Publisher<V>> itemTimeoutIndicator;
 
-        Subscription s;
+        final Function<? super T, ? extends Publisher<?>> itemTimeoutIndicator;
 
-        volatile boolean cancelled;
+        final SequentialDisposable task;
 
-        volatile long index;
+        final AtomicReference<Subscription> upstream;
 
-        final AtomicReference<Disposable> timeout = new AtomicReference<Disposable>();
+        final AtomicLong requested;
 
-        TimeoutSubscriber(Subscriber<? super T> actual,
-                Publisher<U> firstTimeoutIndicator,
-                Function<? super T, ? extends Publisher<V>> itemTimeoutIndicator) {
+        TimeoutSubscriber(Subscriber<? super T> actual, Function<? super T, ? extends Publisher<?>> itemTimeoutIndicator) {
             this.actual = actual;
-            this.firstTimeoutIndicator = firstTimeoutIndicator;
             this.itemTimeoutIndicator = itemTimeoutIndicator;
+            this.task = new SequentialDisposable();
+            this.upstream = new AtomicReference<Subscription>();
+            this.requested = new AtomicLong();
         }
 
         @Override
         public void onSubscribe(Subscription s) {
-            if (!SubscriptionHelper.validate(this.s, s)) {
-                return;
-            }
-            this.s = s;
-
-            if (cancelled) {
-                return;
-            }
-
-            Subscriber<? super T> a = actual;
-
-            Publisher<U> p = firstTimeoutIndicator;
-
-            if (p != null) {
-                TimeoutInnerSubscriber<T, U, V> tis = new TimeoutInnerSubscriber<T, U, V>(this, 0);
-
-                if (timeout.compareAndSet(null, tis)) {
-                    a.onSubscribe(this);
-                    p.subscribe(tis);
-                }
-            } else {
-                a.onSubscribe(this);
-            }
+            SubscriptionHelper.deferredSetOnce(upstream, requested, s);
         }
 
         @Override
         public void onNext(T t) {
-            long idx = index + 1;
-            index = idx;
-
-            actual.onNext(t);
+            long idx = get();
+            if (idx == Long.MAX_VALUE || !compareAndSet(idx, idx + 1)) {
```

Comment:
nit: maybe leave a link to your blog post in comment? So readers could understand the pattern used here

Reply:
Not really keen on this. The PR will reference the blog.

Reply:
Right, my point is not to promote your blog haha

Just want to leave some bread crumbs in the code so one could understand that `MAX_VALUE` is used as a notion of terminal state here, one might confuse it with something related to unbounded request mode

I'd just put 

```java
@see <a href="https://akarnokd.blogspot.hu/2017/09/java-9-flow-api-timing-out-events.html">Operator Design Principles</a>
```

To the class's javadoc.

It's obviously a nit and up2u :)

Reply:
Thanks for the suggestion, but I won't put it there.